### ISSUE TITLE
Unsupported parameters for (ios_command) module - GNS3 -Win10 Linux Emulator

### ISSUE THREAD

**Issue by ajayreddy28390** [ansible/ansible#35697]
##### ISSUE TYPE
 - Bug Report

##### ANSIBLE VERSION

```
ansible 2.4.3.0
  config file = /mnt/c/Users/ajayr/Documents/overall_example/ansible.cfg
  configured module search path = [u'/home/ajayreddy28390/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/ajayreddy28390/.local/lib/python2.7/site-packages/ansible
  executable location = /home/ajayreddy28390/.local/bin/ansible
  python version = 2.7.12 (default, Dec  4 2017, 14:50:18) [GCC 5.4.0 20160609]
```

##### CONFIGURATION
DEFAULT_HOST_LIST(/mnt/c/Users/ajayr/Documents/overall_example/ansible.cfg) = [u'/mnt/c/Users/ajayr/Documents/overall_example/hosts']
DEFAULT_TIMEOUT(/mnt/c/Users/ajayr/Documents/overall_example/ansible.cfg) = 5
DEFAULT_TRANSPORT(/mnt/c/Users/ajayr/Documents/overall_example/ansible.cfg) = paramiko
HOST_KEY_CHECKING(/mnt/c/Users/ajayr/Documents/overall_example/ansible.cfg) = False

##### SUMMARY

I can SSH to routers on GNS3 from Linux Emulator on Win10 PC but when i run Ansible. It throws below error.

```
fatal: [192.168.24.250.xip.io]: FAILED! => {"changed": false, "msg": "Unsupported parameters for (ios_command) module: transport found in provider. Supported parameters include: auth_pass,authorize,host,password,port,ssh_keyfile,timeout,username"}
fatal: [192.168.24.254.xip.io]: FAILED! => {"changed": false, "msg": "Unsupported parameters for (ios_command) module: transport found in provider. Supported parameters include: auth_pass,authorize,host,password,port,ssh_keyfile,timeout,username"}
```

##### EXPECTED RESULTS
Should get command outputs displayed

##### ACTUAL RESULTS

```
<192.168.24.250.xip.io> using connection plugin network_cli
<192.168.24.254.xip.io> using connection plugin network_cli
<192.168.24.250.xip.io> socket_path: /home/ajayreddy28390/.ansible/pc/d19932adfb
<192.168.24.254.xip.io> socket_path: /home/ajayreddy28390/.ansible/pc/0322650089
Using module file /home/ajayreddy28390/.local/lib/python2.7/site-packages/ansible/modules/network/ios/ios_command.py
<192.168.24.250> ESTABLISH LOCAL CONNECTION FOR USER: ajayreddy28390
<192.168.24.250> EXEC /bin/sh -c 'echo ~ && sleep 0'
<192.168.24.250> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /home/ajayreddy28390/.ansible/tmp/ansible-tmp-1517706780.72-205718937085550 `" && echo ansible-tmp-1517706780.72-205718937085550="` echo /home/ajayreddy28390/.ansible/tmp/ansible-tmp-1517706780.72-205718937085550 `" ) && sleep 0'
<192.168.24.250> PUT /tmp/tmp8azYCJ TO /home/ajayreddy28390/.ansible/tmp/ansible-tmp-1517706780.72-205718937085550/ios_command.py
<192.168.24.250> EXEC /bin/sh -c 'chmod u+x /home/ajayreddy28390/.ansible/tmp/ansible-tmp-1517706780.72-205718937085550/ /home/ajayreddy28390/.ansible/tmp/ansible-tmp-1517706780.72-205718937085550/ios_command.py && sleep 0'
Using module file /home/ajayreddy28390/.local/lib/python2.7/site-packages/ansible/modules/network/ios/ios_command.py
<192.168.24.254> ESTABLISH LOCAL CONNECTION FOR USER: ajayreddy28390
<192.168.24.254> EXEC /bin/sh -c 'echo ~ && sleep 0'
<192.168.24.250> EXEC /bin/sh -c '/usr/bin/python /home/ajayreddy28390/.ansible/tmp/ansible-tmp-1517706780.72-205718937085550/ios_command.py; rm -rf "/home/ajayreddy28390/.ansible/tmp/ansible-tmp-1517706780.72-205718937085550/" > /dev/null 2>&1 && sleep 0'
<192.168.24.254> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /home/ajayreddy28390/.ansible/tmp/ansible-tmp-1517706781.52-136227427656800 `" && echo ansible-tmp-1517706781.52-136227427656800="` echo /home/ajayreddy28390/.ansible/tmp/ansible-tmp-1517706781.52-136227427656800 `" ) && sleep 0'
<192.168.24.254> PUT /tmp/tmpD_se92 TO /home/ajayreddy28390/.ansible/tmp/ansible-tmp-1517706781.52-136227427656800/ios_command.py
<192.168.24.254> EXEC /bin/sh -c 'chmod u+x /home/ajayreddy28390/.ansible/tmp/ansible-tmp-1517706781.52-136227427656800/ /home/ajayreddy28390/.ansible/tmp/ansible-tmp-1517706781.52-136227427656800/ios_command.py && sleep 0'
<192.168.24.254> EXEC /bin/sh -c '/usr/bin/python /home/ajayreddy28390/.ansible/tmp/ansible-tmp-1517706781.52-136227427656800/ios_command.py; rm -rf "/home/ajayreddy28390/.ansible/tmp/ansible-tmp-1517706781.52-136227427656800/" > /dev/null 2>&1 && sleep 0'
fatal: [192.168.24.250.xip.io]: FAILED! => {
    "changed": false,
    "invocation": {
        "module_args": {
            "auth_pass": null,
            "authorize": null,
            "commands": [
                "show version"
            ],
            "host": null,
            "interval": 1,
            "match": "all",
            "password": null,
            "port": null,
            "provider": {
                "auth_pass": null,
                "authorize": null,
                "host": "192.168.24.250.xip.io",
                "password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
                "port": null,
                "ssh_keyfile": null,
                "timeout": null,
                "transport": "cli",
                "username": "cisco"
            },
            "retries": 10,
            "ssh_keyfile": null,
            "timeout": null,
            "username": null,
            "wait_for": null
        }
    },
    "msg": "Unsupported parameters for (ios_command) module: transport found in provider. Supported parameters include: auth_pass,authorize,host,password,port,ssh_keyfile,timeout,username"
}
fatal: [192.168.24.254.xip.io]: FAILED! => {
    "changed": false,
    "invocation": {
        "module_args": {
            "auth_pass": null,
            "authorize": null,
            "commands": [
                "show version"
            ],
            "host": null,
            "interval": 1,
            "match": "all",
            "password": null,
            "port": null,
            "provider": {
                "auth_pass": null,
                "authorize": null,
                "host": "192.168.24.254.xip.io",
                "password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
                "port": null,
                "ssh_keyfile": null,
                "timeout": null,
                "transport": "cli",
                "username": "cisco"
            },
            "retries": 10,
            "ssh_keyfile": null,
            "timeout": null,
            "username": null,
            "wait_for": null
        }
    },
    "msg": "Unsupported parameters for (ios_command) module: transport found in provider. Supported parameters include: auth_pass,authorize,host,password,port,ssh_keyfile,timeout,username"
}
        to retry, use: --limit @/mnt/c/Users/ajayr/Documents/overall_example/IOS_test.retry

PLAY RECAP **********************************************************************************************************************************************************************************************************************************
192.168.24.250.xip.io      : ok=0    changed=0    unreachable=0    failed=1
192.168.24.254.xip.io      : ok=0    changed=0    unreachable=0    failed=1
```

**Comment by Akasurde** [comment]
@ajayreddy28390 Thanks for reporting this issue. Could you please provide relevant playbook used for this run ?

needs_info

**Comment by ajayreddy28390** [comment]
Attached relevant files and router configs.
I can SSH from Linux emulator to devices on GNS3. So, credentials and SSH version are good. I did wireshark capture, i see no issues for SSH traffic.
[Ansible_test1_configs.zip](https://github.com/ansible/ansible/files/1692778/Ansible_test1_configs.zip)

[overall_example.zip](https://github.com/ansible/ansible/files/1692777/overall_example.zip)

**Comment by trishnaguha** [comment]
@ajayreddy28390 transport isn't an option that ios module supports. Please check the doc http://docs.ansible.com/ansible/latest/ios_command_module.html. 
The playbook will work after transport is removed from the provider spec.

**Comment by mandar301190** [comment]
I have that kind of same problem. I can SSH to all routers and switches successfully. But when trying to run ansible playbook to configure ospf then it shows error as "changed": false, "msg": "Unsupported parameters for (ios_config) module: transport found in provider' 

Hereby I have attached my script:----
- name: Manage GNS3 devices
  gather_facts: false
  connection: local

  vars:
    cli: 
      username: mandar 
      password: cisco
      transport: cli

  tasks: 
    - name: enable ospf 
      ios_config: 
        provider: "{{ cli }}"
        authorize: yes
        parents: router ospf 1
        lines: 
          - network 0.0.0.0 255.255.255.255 area 0

      register: print_output

    -  debug: var=print_output

**Comment by trishnaguha** [comment]
@mandar301190 Please do not use `transport` for ios modules. It is not supported for ios.

**Comment by mandar301190** [comment]
Alright. But even if I remove transport then what changes I should make in
task section?

**Comment by trishnaguha** [comment]
@mandar301190 no changes required, just remove transport from vars cli.

**Comment by mandar301190** [comment]
Yes I had done exactly the same way but it gives me trace back error. I
don’t know whether the issue is wid Ansible installed version or something
else.

**Comment by trishnaguha** [comment]
@mandar301190 Can you share the traceback and exact playbook you are using?
You will not have the same traceback posted here if you remove transport. Thanks.

**Comment by mandar301190** [comment]
This is my output after removing transport module. I am currently using ansible 2.5.2.
Result:-
PLAY [Manage GNS3 devices] *****************************************************

TASK [enable ospf] *************************************************************
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: ansible.module_utils.connection.ConnectionError: socket_path does not exist or cannot be found. Please check https://docs.ansible.com/ansible/latest/network/user_guide/network_debug_troubleshooting.html#category-socket-path-issue
fatal: [R1]: FAILED! => {"changed": false, "module_stderr": "Traceback (most recent call last):\n  File \"/tmp/ansible_O291Bs/ansible_module_ios_config.py\", line 561, in <module>\n    main()\n  File \"/tmp/ansible_O291Bs/ansible_module_ios_config.py\", line 490, in main\n    load_config(module, commands)\n  File \"/tmp/ansible_O291Bs/ansible_modlib.zip/ansible/module_utils/network/ios/ios.py\", line 168, in load_config\n  File \"/tmp/ansible_O291Bs/ansible_modlib.zip/ansible/module_utils/connection.py\", line 168, in __rpc__\n  File \"/tmp/ansible_O291Bs/ansible_modlib.zip/ansible/module_utils/connection.py\", line 142, in _exec_jsonrpc\nansible.module_utils.connection.ConnectionError: socket_path does not exist or cannot be found. Please check https://docs.ansible.com/ansible/latest/network/user_guide/network_debug_troubleshooting.html#category-socket-path-issue\n", "module_stdout": "", "msg": "MODULE FAILURE", "rc": 1}
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: ansible.module_utils.connection.ConnectionError: socket_path does not exist or cannot be found. Please check https://docs.ansible.com/ansible/latest/network/user_guide/network_debug_troubleshooting.html#category-socket-path-issue
fatal: [R2]: FAILED! => {"changed": false, "module_stderr": "Traceback (most recent call last):\n  File \"/tmp/ansible_m33EAF/ansible_module_ios_config.py\", line 561, in <module>\n    main()\n  File \"/tmp/ansible_m33EAF/ansible_module_ios_config.py\", line 490, in main\n    load_config(module, commands)\n  File \"/tmp/ansible_m33EAF/ansible_modlib.zip/ansible/module_utils/network/ios/ios.py\", line 168, in load_config\n  File \"/tmp/ansible_m33EAF/ansible_modlib.zip/ansible/module_utils/connection.py\", line 168, in __rpc__\n  File \"/tmp/ansible_m33EAF/ansible_modlib.zip/ansible/module_utils/connection.py\", line 142, in _exec_jsonrpc\nansible.module_utils.connection.ConnectionError: socket_path does not exist or cannot be found. Please check https://docs.ansible.com/ansible/latest/network/user_guide/network_debug_troubleshooting.html#category-socket-path-issue\n", "module_stdout": "", "msg": "MODULE FAILURE", "rc": 1}
	to retry, use: --limit @/root/myansiblee1.retry

PLAY RECAP *********************************************************************
R1                         : ok=0    changed=0    unreachable=0    failed=1   
R2                         : ok=0    changed=0    unreachable=0    failed=1

**Comment by mandar301190** [comment]
This is my playbook script. Just for initial checking purpose I am configuring simple ospf protocol on two routers. 

Script:

---
- name: Manage GNS3 devices
  hosts: R1, R2
  gather_facts: false
  connection: local

  vars:
   cli:
    username: mandar
    password: cisco       
   

  tasks: 
    - name: enable ospf 
      ios_config: 
        provider: "{{ cli }}"
        authorize: yes
        parents: router ospf 1
        lines: 
          - network 0.0.0.0 255.255.255.255 area 0

      register: print_output

    -  debug: var=print_output

**Comment by trishnaguha** [comment]
@mandar301190 Please try the troubleshooting guide https://docs.ansible.com/ansible/latest/network/user_guide/network_debug_troubleshooting.html#category-socket-path-issue. 
The documentation shows how to do debug enabling logging https://docs.ansible.com/ansible/latest/network/user_guide/network_debug_troubleshooting.html#enabling-networking-logging-and-how-to-read-the-logfile.
This can be related to timeout: If see timeout issue in your debug log please try this out https://docs.ansible.com/ansible/latest/network/user_guide/network_debug_troubleshooting.html#timeout-issues

**Comment by mandar301190** [comment]
Hey 
I just changed version of ansible and accordingly have to make in script. More on timeout value change worked. So that you for that. 
But once configuration is done on switch or router, when I execute same command again within few seconds, it gives me traceback error. Then have to try consistently by same command and it works. Any idea why this could be an issue?

**Comment by mandar301190** [comment]
Trishna,
This is another issue with linux container that I am working on. Issue seems simple but frustrating.
I am not able to SSH to any one of the containers using ansible, where as I have already configured SSH key successfully and agent able to authenticate. In my opinion public key is some what issue. Please have a look on following output and let me know your feedback.

Output:
root@ubuntu:~# ssh-add -L

ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQDWIyLseA3+PQfZoXVp9tirexN8d5/se2UZf81tPNu+4JVH7jScejGM8WmDTM//BENOMpKJVqqMr9L/e2Fwo4wqWJjihLZUBTcADtcFm/GqSAxXGIUjEsCAyB2aLvlk+52BCtEI1v5imZkeWC2+l9gX27gFcUa9GI3+0fcNNNYz6FbgzRkjiCc8sLpPwBXDONto5qsrGyTODrD9Kysy3lhX4WV1k2gDYe1oQOOmbgvTsqs7tGSEBoC52do4EbR0W+YKW5B9OitGalcCtqQ/2DDMQ7t0fq+Ji9rFVf/ScuEf8llQfo03HTkhByJh1O+Y3sYiPS3+3QICIgzVD7C4zrdV /root/.ssh/id_rsa

root@ubuntu:~# ssh-add ~/.ssh/id_rsa

Identity added: /root/.ssh/id_rsa (/root/.ssh/id_rsa)

root@ubuntu:~# ansible 10.0.3.201 -m ping -u root

10.0.3.201 | UNREACHABLE! => {

    "changed": false, 

    "msg": "Failed to connect to the host via ssh: Warning: Permanently added '10.0.3.201' (ECDSA) to the list of known hosts.\r\nroot@10.0.3.201: Permission denied (publickey,password).\r\n", 

    "unreachable": true
